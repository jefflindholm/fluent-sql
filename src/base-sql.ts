export interface BaseColumn {
  ColumnName: string | undefined;
  Alias: string | undefined;
  Literal: string | undefined;
  [key: string]: any;
}
export interface BaseTable {
  TableName: string | null;
  Alias: string;
  Columns: BaseColumn[];
  Schema: string | null;
  [key: string]: any;
  getTable(query?: BaseQuery | null): string;
}
export interface BaseWhere {
  Column: BaseColumn | null;
  Op: string | null;
  Value: any;
  Wheres: BaseWhere[];
}
export enum eEscapeLevels {
  tableAlias = 'table-alias',
  columnAlias = 'column-alias',
  tableName = 'table-name',
  columnName = 'column-name',
}
export interface BaseQuery {
  sqlEscape(str: string, level: eEscapeLevels): string;
}
