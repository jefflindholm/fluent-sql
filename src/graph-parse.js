const isNumeric = (val) => parseFloat(val) == val;

function buildParams(query) {
    const results = [];

    // handle params p:v,p1:v1,...pn:vn
    // v = literal or v = <variable> or v = &ref
    const params = query.split(',');
    params.forEach(p => {
        const pieces = p.split(':');
        const param = {
            type: 'Argument',
            name: pieces[0],
        };
        if (pieces[1][0] === '<') {
            param.value = {
                type: 'Variable',
                name: pieces[1].replace(/[<>]/g, ''),
            };
        } else if (pieces[1][0] === '&') {
            const name = pieces[1].substring(1);
            param.value = {
                type: 'Reference',
                name: isNumeric(name) ? parseFloat(name) : name,
            };
        } else {
            let value = '';
            if (pieces[1][0] === '"') {
                value = pieces[1].replace(/"/g, '');
            } else if (pieces[1] === 'true' || pieces[1] === 'false') {
                value = !!pieces[1];
            } else if (pieces[1] === 'null') {
                value = null;
            } else {
                value = parseFloat(pieces[1]);
            }
            param.value = {
                type: 'Literal',
                value,
            }
        }
        results.push(param);
    });
    return results;
}

const newField = {
    type: 'Field',
    name: '',
    alias: null,
    params: [],
    fields: []
};

function parse(baseQuery) {
    const query = baseQuery.replace(/\sas\s/, '*as*').replace(/\s/g, '');

    const fields = [];
    let field = null;
    for (let i = 0; i < query.length; i++) {
        let c = query[i];

        // start of sub field list
        if (c === '{') {
            if (field ) {
                const result = parse(query.substr(i));
                field.fields = result.fields;
                i += result.end;
            } else {
                field = Object.assign({}, newField);
                fields.push(field);
            }
            continue;
        }
        // the sub fields are done
        if (c === '}') {
            return {
                fields,
                end: i,
            };
        }

        // Handle foo as bar
        if (c === '*') {
            i += 4;
            field.alias = '';
            while (query[i] !== ',' && query[i] !== '(' && query[i] !== '{') {
                field.alias += query[i++];
            }
            i = i - 1;
            continue;
        }

        // our field has some parameters
        if (c === '(') {
            // pull everything from between ( and )
            let end = i;
            while (query[end] !== ')') end++;
            const params = query.substr(i + 1, end - (i + 1));
            i = end;
            field.params = buildParams(params);
            continue;
        }

        // our field is complete, create a new field
        if (c === ',') {
            field = Object.assign({}, newField);
            fields.push(field);
            continue;
        }

        field.name += c;
    }
    return { fields, end: query.length };
}
export default (query) => {
    return {
        type: 'Query',
        fields: parse(query).fields,
    }
}
