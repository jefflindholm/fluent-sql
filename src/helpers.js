export function processArgs(func, ...args) {
    for (let i = 0; i < args.length; i++) {
        if (Array.isArray(args[i])) {
            for (let j = 0; j < args[i].length; j++) {
                func(args[i][j]);
            }
        } else {
            func(args[i]);
        }
    }
}
export class SqlError extends Error {
    constructor(loc, msg) {
        super(msg);
        this.location = loc;
        this.message = msg;
    }
}
