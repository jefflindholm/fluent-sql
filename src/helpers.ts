export function processArgs(func: any, ...args: any) {
    for (let i = 0; i < args.length; i++) {
        if (Array.isArray(args[i])) {
            for (let j = 0; j < args[i].length; j++) {
                func(args[i][j]);
            }
        } else {
            func(args[i]);
        }
    }
}
export class SqlError {
    constructor(loc: string, msg: string) {
        // super(msg);
        this.location = loc;
        this.message = msg;
    }
  location: string;
  message: string;
}
