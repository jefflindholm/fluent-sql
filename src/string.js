function toCamel() {
    return this.replace(/[\-_][a-z]/g, (m) => {
        return m.toUpperCase().replace(/[\-_]/, '');
    });
};
function trim()  {
    return this.replace(/^\s+|\s+$/g, '');
};
function toDashCase()  {
    let result = this.replace(/[A-Z]/g, (m) => {
        return `-${m.toLowerCase()}`;
    });
    result = result.replace(/_/g, '-')
    return (result[0] === '-') ? result.substring(1) : result;
};
function toSnakeCase()  {
    let result = this.replace(/[A-Z]/g, (m) => {
        return `_${m.toLowerCase()}`;
    });
    result = result.replace(/\-/g, '_')
    return (result[0] === '_') ? result.substring(1) : result;
};
function capitalizeFirst()  {
    return this.charAt(0).toUpperCase() + this.slice(1);
};
function toPascal()  {
    return this.toCamel().capitalizeFirst();
};
function contains (s) {
    return this.indexOf(s) > -1;
};

(function() {
    'use strict';
    const stringFunctions = {
        toCamel,
        trim,
        toDashCase,
        toSnakeCase,
        capitalizeFirst,
        toPascal,
        contains,
    };

    for (const key in stringFunctions) {
        if (String.prototype.hasOwnProperty(key)) {
            continue;
        }
        Object.defineProperty(String.prototype, key, {
            value: stringFunctions[key],
            writable: true,
        });
    }

})();

export default {
    toCamel,
    trim,
    toDashCase,
    toSnakeCase,
    capitalizeFirst,
    toPascal,
    contains,
};
