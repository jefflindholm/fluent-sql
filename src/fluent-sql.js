"use strict";

import './string.js';
import sliced from 'sliced';
import util from 'util';
import {sprintf} from 'sprintf-js';

import SqlQuery from './sql-query';

export { default as  SqlBuilder } from './sql-builder';
export { default as  SqlColumn } from './sql-column';
export { default as  SqlJoin } from './sql-join';
export { default as  SqlOrder } from './sql-order';
export { default as  SqlTable } from './sql-table';
export { default as  SqlWhere } from './sql-where';
export { default as  SqlQuery } from './sql-query';

if (!String.prototype.sqlEscape) {
    (function () {
        'use strict';
        var sqlEscape = function (sqlQuery, level) {

            if (!sqlQuery || !sqlQuery.sqlEscape || typeof sqlQuery.sqlEscape !== "function") {
                sqlQuery = new SqlQuery();
            }
            return sqlQuery.sqlEscape(this, level);
        }
        String.prototype.sqlEscape = sqlEscape;
    })();
}
