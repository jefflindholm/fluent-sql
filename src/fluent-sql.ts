import './string.js';

import SqlQuery from './sql-query';

export { default as SqlBuilder } from './sql-builder';
export { default as SqlColumn } from './sql-column';
export { default as SqlJoin } from './sql-join';
export { default as SqlOrder } from './sql-order';
export { default as SqlTable } from './sql-table';
export { default as SqlWhere } from './sql-where';
export { default as SqlQuery } from './sql-query';
export { SqlError } from './helpers';
export {
    setDefaultOptions,
    getDefaultOptions,
    setPostgres,
    setSqlServer,
    postgresOptions,
    sqlServerOptions
} from './sql-query';
