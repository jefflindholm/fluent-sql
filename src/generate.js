#!/usr/bin/env node
const { commandLineParser } = require('simple-db-migrate/command-line');
require('./string.js');

let db;

async function configure(newConfig) {
  config = {
    dialect: process.env.MIGRATE_DIALECT || 'sqlite',
    database: process.env.MIGRATE_DATABASE || 'db.sqlite',
    verbose: false,
    debug: false,
    ...newConfig
  };

  const dialect = config.dialect.toLowerCase();
  db = require(`./generate-lib/dialect-${dialect}`);

  await db.configure(config);
}

const options = [
  { name: 'verbose', alias: 'v', type: Boolean },
  { name: 'dialect', alias: 'd', options: ['pg', 'sqlite'] },
  { name: 'database', alias: 'db', type: String },
  { name: 'user', alias: 'u', type: String },
  { name: 'password', alias: 'p', type: String },
  { name: 'host', alias: 'h', type: String },
  { name: 'port', type: Number, numeric: 'int' },
];

const db_config = commandLineParser(options);

(async () => {
  await configure(db_config);
  const tables = await db.getTables();
  await tables.forEach(async t => {
    const columns = await db.getColumns(t.name);
    let sqlTable = `const ${t.name.toCamel()} = new SqlTable('${t.name}', [\n`
    columns.forEach(c => {
      const col = `  {ColumnName: '${c.name}'},`
      sqlTable += `${col}${' '.repeat(30 - col.length)}// type: '${c.type}'\n`
    });
    sqlTable += ']);\n\n';
    console.log(sqlTable);
  });
})();
