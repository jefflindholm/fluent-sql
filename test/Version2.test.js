/* global describe it */
import '../src/string';
import { SqlQuery } from '../src/fluent-sql.js';
import { SqlTable } from '../src/fluent-sql.js';
import { SqlColumn } from '../src/fluent-sql.js';
//import {SqlWhere} from '../src/fluent-sql.js';
//import {SqlOrder} from '../src/fluent-sql.js';
import { SqlJoin } from '../src/fluent-sql.js';
import { SqlBuilder } from '../src/fluent-sql.js';
import { setDefaultOptions, getDefaultOptions, postgresOptions, sqlServerOptions } from '../src/fluent-sql.js';

describe('version2 tests', () => {
  const businessColumns = [{ ColumnName: 'id' }, { ColumnName: 'business_name' }, { ColumnName: 'tax_id' }];
  const business = new SqlTable({ TableName: 'business', columns: businessColumns });

  describe('Postgres options', () => {
    it('should build a query with $# for params for simple sql', () => {
      setDefaultOptions(postgresOptions);
      const query = new SqlQuery()
        .from(business)
        .select(business.star())
        .where(business.businessName.like('foo'));
      const sql = query.genSql();
      expect(sql.fetchSql).toBe(
        'SELECT\n"business".id as "id",\n"business".business_name as "businessName",\n"business".tax_id as "taxId"\nFROM\nbusiness as "business"\nWHERE "business".business_name like ($1)',
      );
      expect(sql.values[0]).toBe('%foo%');
    });
    it('should build a query with $# for params for paging', () => {
      setDefaultOptions(postgresOptions);
      const query = new SqlQuery()
        .from(business)
        .pageSize(1)
        .select(business.star())
        .where(business.businessName.like('foo'));
      const sql = query.genSql();
      expect(sql.fetchSql).toBe(
        'SELECT * FROM (' +
          '\nSELECT *, row_number() OVER (ORDER BY "id") as Paging_RowNumber FROM (' +
          '\nSELECT' +
          '\n"business".id as "id",' +
          '\n"business".business_name as "businessName",' +
          '\n"business".tax_id as "taxId"' +
          '\nFROM\nbusiness as "business"' +
          '\nWHERE "business".business_name like ($1)' +
          '\n) base_query' +
          '\n) as detail_query WHERE Paging_RowNumber BETWEEN 0 AND 1',
      );
      expect(sql.values[0]).toBe('%foo%');
    });
    it('should build a query with LIMIT not TOP', () => {
      setDefaultOptions(postgresOptions);
      const query = new SqlQuery()
        .from(business)
        .top(10)
        .select(business.star())
        .orderBy(business.id)
        .where(business.businessName.like('foo'));
      const sql = query.genSql();
      expect(sql.fetchSql).toBe(
        'SELECT' +
          '\n"business".id as "id",' +
          '\n"business".business_name as "businessName",' +
          '\n"business".tax_id as "taxId"' +
          '\nFROM' +
          '\nbusiness as "business"' +
          '\nWHERE "business".business_name like ($1)' +
          '\nORDER BY "business".id ASC' +
          '\nLIMIT 10',
      );
      expect(sql.values[0]).toBe('%foo%');
    });
    it('should generate update with $#', () => {
      setDefaultOptions(postgresOptions);
      const update = SqlBuilder.update(business, { businessName: 'Test', id: 1 });
      expect(update).toMatchObject({
        sql: 'UPDATE business SET business_name = $2 WHERE id = $1',
        values: [1, 'Test'],
        hasEncrypted: false,
      });
    });
    it('should generate delete with $#', () => {
      const del = SqlBuilder.delete(business, { businessName: 'Test', id: 1 });
      expect(del).toMatchObject({
        sql: 'DELETE FROM business WHERE business_name = $2 AND id = $1',
        values: [1, 'Test'],
        hasEncrypted: false,
      });
    });
    it('should generate insert with $#', () => {
      const insert = SqlBuilder.insert(business, { businessName: 'Test', id: 1 });
      expect(insert).toMatchObject({
        sql: 'INSERT INTO business (business_name) VALUES ($1)',
        values: ['Test'],
        hasEncrypted: false,
      });
    });
    it('should generate insert with $# and new id', () => {
      const insert = SqlBuilder.insert(business, { businessName: 'Test', id: 1 }, 4000);
      expect(insert).toMatchObject({
        sql: 'INSERT INTO business (business_name, id) VALUES ($1, $2)',
        values: ['Test', 4000],
        hasEncrypted: false,
      });
    });
  });

  describe('missing tests', () => {
    describe('missing SqlJoin', () => {
      it('should throw execption when created with something not a SqlColumn', () => {
        expect(() => new SqlJoin({})).toThrowErrorMatchingSnapshot();
      });
      it('should throw execption when using a non SqlColumn', () => {
        const join = new SqlJoin(business.businessName);
        expect(() => join.using({})).toThrowErrorMatchingSnapshot();
      });
    });
    describe('missing SqlColumn', () => {
      it('should allow a starts for a column where', () => {
        setDefaultOptions(postgresOptions);
        const query = new SqlQuery()
          .from(business)
          .select(business.star())
          .where(business.businessName.starts('foo'));
        const sql = query.genSql();
        expect(sql.fetchSql).toBe(
          'SELECT\n"business".id as "id",\n"business".business_name as "businessName",\n"business".tax_id as "taxId"\nFROM\nbusiness as "business"\nWHERE "business".business_name like ($1)',
        );
        expect(sql.values[0]).toBe('foo%');
      });
      it('should allow a ends for a column where', () => {
        setDefaultOptions(postgresOptions);
        const query = new SqlQuery()
          .from(business)
          .select(business.star())
          .where(business.businessName.ends('foo'));
        const sql = query.genSql();
        expect(sql.fetchSql).toBe(
          'SELECT\n"business".id as "id",\n"business".business_name as "businessName",\n"business".tax_id as "taxId"\nFROM\nbusiness as "business"\nWHERE "business".business_name like ($1)',
        );
        expect(sql.values[0]).toBe('%foo');
      });
    });
    describe('missing SqlQuery', () => {
      it('should throw an execption if there is no FROM', () => {
        const query = new SqlQuery().select(business.star());
        expect(() => query.genSql()).toThrowErrorMatchingSnapshot();
      });
    });
  });
});
