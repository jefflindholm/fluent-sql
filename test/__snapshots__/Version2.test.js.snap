// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`version2 tests missing tests missing SqlJoin should throw execption when created with something not a SqlColumn 1`] = `"trying to join on something not a SqlColumn"`;

exports[`version2 tests missing tests missing SqlJoin should throw execption when using a non SqlColumn 1`] = `"trying to join on something not a SqlColumn"`;

exports[`version2 tests missing tests missing SqlQuery should throw an execption if there is no FROM 1`] = `"No FROM in query"`;
